core/src/blockchain_rpc/client.rs

```diff
@@ -132,19 +145,36 @@ impl BlockchainRpcClient {
 
         let targets = client.get_broadcast_targets();
         let request = Request::from_tl(ExternalMessage { data: message });
-
         let mut futures = FuturesUnordered::new();
+
+        // we wait for all the responses to come back but cap them at `broadcast_timeout_upper_bound`
+        // all peers timeouts are calculated based on p90 of the previous responses time weighted average
+        // This will make broadcast timeout to be adaptive based on the network conditions
         for validator in targets.as_ref() {
             let client = client.clone();
             let validator = validator.clone();
             let request = request.clone();
-            futures.push(JoinTask::new(async move {
-                client.send_to_validator(validator, request).await
+            let this = self.clone();
+            // we are not using `JoinTask` here because we want to measure the time taken by the broadcast
+            futures.push(tokio::spawn(async move {
+                let start = Instant::now();
+                let res = tokio::time::timeout(
+                    this.inner.broadcast_timeout_upper_bound,
+                    client.send_to_validator(validator, request),
+                )
+                .await;
+                this.inner
+                    .response_tracker
+                    .lock()
+                    .append(start.elapsed().as_millis() as i64);
+                res
             }));
         }
 
-        tokio::time::timeout(self.inner.broadcast_timeout, async {
-            while let Some(res) = futures.next().await {
+        let timeout = self.broadcast_timeout();
+        tokio::time::timeout(timeout, async {
+            // inner task timeout won't happen because outer task timeout is always <= inner task timeout
```

Comment:
Then why is the inner timeout even needed?

Reply:
If there is are neighbours that are "dying" and are slow to respond, this method can spawn too many broadcasts. For that case you might need to track a number of running futures (just an atomic) and consume the remaining of futures after the timeout and cancel them while the number is greater than some threshold